domain/application/state/application.go

```diff
@@ -3466,3 +3472,61 @@ func (st *State) precheckUpgradeRelation(ctx context.Context, tx *sqlair.TX, app
 	}
 	return nil
 }
+
+func (st *State) refreshApplicationConfig(ctx context.Context, tx *sqlair.TX, appIdent applicationID, charmIdent charmID) error {
+	charmConfig, err := st.getCharmConfig(ctx, tx, charmIdent)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	decodedCharmConfig, err := application.DecodeConfig(charmConfig)
+	if err != nil {
+		return errors.Errorf("decoding charm config: %w", err)
+	}
+
+	applicationConfig, err := st.getApplicationConfig(ctx, tx, appIdent)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	decodedApplicationConfig := make(internalcharm.Config, len(applicationConfig))
+	for _, v := range applicationConfig {
+		decodedApplicationConfig[v.Key] = v.Value
+	}
```

Comment:
It is unfortunate that these types have leaked their way into the state layer. But it is necessary to keep transactionality